packages/mesos/extra/dcos-drain-mesos-agent-public.service

```diff
@@ -0,0 +1,9 @@
+[Unit]
+Description=Drain Mesos Agent: Stop Mesos tasks on shutdown/reboot
+Before=shutdown.target reboot.target halt.target
+After=dcos-mesos-slave-public.service
+
+[Service]
+Type=oneshot
+RemainAfterExit=true
+ExecStop=/usr/bin/systemctl kill -s SIGUSR1 dcos-mesos-slave-public.service
```

Comment:
I don't know systemd enough, but what's the guarantee that the agent *finishes* the draining before the host is rebooted? Does the "After" statement ensure that until the dcos agent service exits, the agent host reboot/shutdown is blocked?

Reply:
I myself had to spend some time in Systemd man for this. So its understandable to have questions around this. 

I am currently not doing anything to make sure the agent has actually finished draining. Is there a way to do determine programmatically that the agent has finished draining. If yes then I can wrap the SIGUSR1 in a script to make sure this service blocks reboot until the agent is drained.  

After=dcos-mesos-slave.service is there to ensure ensure that this is stopped before dcos-mesos-slave.service (stopping order is reverse of starting)

Reply:
I think you can wait until the agent process exits to ensure draining is complete. But if for whatever reason the agent doesn't exit by itself (e.g., bug in draining path), then we would be stuck. The script you are proposing or this systemd unit should probably have a timeout as a fail safe.

Reply:
Also, I'm wondering if this drain should be part of the `service stop` workflow. Right now, we tell users in our docs that they need to drain via SIGUSR1 and then do `service stop`. Would be nice if they just do `stop` and the draining happens. Unless, there are cases where someone would want to stop the agent process but wants to keep the tasks running. Can `stop` take arguments (e.g., --drain)? I'm just spitballing here. I just find it weird that we expose users directly to the SIGUSR1, instead of a `drain` systemd command.

Reply:
My very first though when I was playing around with this pre DCOS_OSS-3672 was to just add an ExecStop for the agent service. But then I recalled that we have used systemctl restart dcos-mesos-slave during debugging and have come to expect restarting mesos agent service to be a non-disruptive process and for good reason : Mesos agent service restart have always been non-disruptive long before DC/OS

I did not want to anything very fancy with systemd to I thought doing a separate unit dedicated for draining would be the best approach. I will google a bit to see if there is something else systemd has that we can use. The challenge is going to be the limitation that we only want to drain on system reboot/halt and not on service restart (which is divergent from behaviour a typical systemd service like HTTP server would want to do)

Reply:
Great point. Another reason not to drain the agent every time we stop it is that we expect DC/OS upgrades to not kill tasks.